pkg/planner/core/tiflash_predicate_push_down.go

```diff
@@ -0,0 +1,371 @@
+func isPredicateSimpleCompare(cond expression.Expression) bool {
+	if sf, ok := cond.(*expression.ScalarFunction); ok {
+		switch sf.FuncName.L {
+		case ast.EQ, ast.GE, ast.LE, ast.LT, ast.GT, ast.In:
+			return true
+		case ast.LogicAnd, ast.LogicOr:
+			return isPredicateSimpleCompare(sf.GetArgs()[0]) && isPredicateSimpleCompare(sf.GetArgs()[1])
+		case ast.UnaryNot:
+			return isPredicateSimpleCompare(sf.GetArgs()[0])
+		}
+	}
+	return false
+}
+
+// handleTiFlashPredicatePushDown is used to handle the TiFlash predicate push down.
+// 1. Whether to use the inverted index.
+// 2. Whether to push down the conditions to the table scan.
+func handleTiFlashPredicatePushDown(pctx base.PlanContext, ts *PhysicalTableScan, indexHints []*ast.IndexHint) {
+	// Consider use index hints
+	indexMap := make(map[string]int, len(ts.Table.Indices))
+	for _, hint := range indexHints {
+		if hint.HintScope != ast.HintForScan {
+			continue
+		}
+
+		switch hint.HintType {
+		case ast.HintUse:
+			for _, name := range hint.IndexNames {
+				if indexMap[name.L] != math.MaxInt {
+					indexMap[name.L]++
+				}
+			}
+		case ast.HintIgnore:
+			for _, name := range hint.IndexNames {
+				if indexMap[name.L] != math.MaxInt {
+					indexMap[name.L]--
+				}
+			}
+		case ast.HintForce:
+			for _, name := range hint.IndexNames {
+				indexMap[name.L] = math.MaxInt
+			}
+		default:
+			continue
+		}
+	}
+
+	indexedColumnNameToIndexInfoMap := make(map[string]*model.IndexInfo, len(ts.Table.Indices))
+	for _, index := range ts.Table.Indices {
+		if index.State == model.StatePublic && index.InvertedInfo != nil {
+			if len(indexHints) > 0 && indexMap[index.Name.L] <= 0 {
+				continue
+			}
+			// inverted index only support one column
+			indexedColumnNameToIndexInfoMap[index.Columns[0].Name.L] = index
+		}
+	}
+
+	selectedColumns := make(map[string]bool, len(indexedColumnNameToIndexInfoMap))
+	selectedConditions := make([]expression.Expression, 0, len(ts.filterCondition))
+	for _, cond := range ts.filterCondition {
+		// 1. The predicate only contains >, >=, =, !=, <=, <, in.
+		if !isPredicateSimpleCompare(cond) {
+			continue
+		}
+		// 2. The columns of predicate all have inverted index.
+		allHaveInvertedIndex := true
+		columns := expression.ExtractColumns(cond)
+		columnNames := make([]string, 0, len(columns))
+		for _, col := range columns {
+			parts := strings.Split(col.OrigName, ".")
+			columnNames = append(columnNames, strings.ToLower(parts[len(parts)-1]))
+		}
+		for _, colName := range columnNames {
+			if _, ok := indexedColumnNameToIndexInfoMap[colName]; !ok {
+				allHaveInvertedIndex = false
+				break
+			}
+		}
+		if !allHaveInvertedIndex {
+			continue
+		}
+		// 3. The selectivity of the predicate is less than 60%.
+		selectivity, _, err := cardinality.Selectivity(ts.SCtx(), ts.tblColHists, []expression.Expression{cond}, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("calculate selectivity failed", zap.Error(err))
```

Comment:
Please don't use ```debug``` log in the planner.

Reply:
https://github.com/pingcap/tidb/blob/c0176459daeed57ca628687fbb42529b53b68070/pkg/planner/core/find_best_task.go#L1918-L1922

I find printing a debug log when meets `calculate selectivity failed` is a common behavior in current code base.

Reply:
done